src/qililab/result/experiment_results.py

```diff
@@ -0,0 +1,273 @@
+# Copyright 2023 Qilimanjaro Quantum Tech
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# mypy: disable-error-code="attr-defined"
+from datetime import datetime
+from typing import Any
+
+import h5py
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class ExperimentResults:
+    """Provides methods to access the experiment results stored in an HDF5 file."""
+
+    QPROGRAMS_PATH = "qprograms"
+    MEASUREMENTS_PATH = "measurements"
+    RESULTS_PATH = "results"
+    VARIABLES_PATH = "variables"
+    EXPERIMENT_PATH = "experiment"
+    PLATFORM_PATH = "platform"
+    EXECUTED_AT_PATH = "executed_at"
+    EXECUTION_TIME_PATH = "execution_time"
+
+    def __init__(self, path: str):
+        """Initializes the ExperimentResults instance.
+
+        Args:
+            path (str): The file path to the HDF5 results file.
+        """
+        self.path = path
+        self.data: dict[tuple[str, str], Any] = {}  # To hold links to the data of the results for in-memory access
+        self.dimensions: dict[
+            tuple[str, str], Any
+        ] = {}  # To hold links to dimensions of the results for in-memory access
+        self._file: h5py.File
+
+    def __enter__(self):
+        """Opens the HDF5 file for reading.
+
+        Returns:
+            ExperimentResults: The ExperimentResults instance.
+        """
+        self._file = h5py.File(self.path, mode="r")
+
+        # Prepare access to each results dataset and its dimensions
+        for qprogram_name in self._file[ExperimentResults.QPROGRAMS_PATH]:
+            qprogram_data = self._file[f"{ExperimentResults.QPROGRAMS_PATH}/{qprogram_name}"]
+            for measurement_name in qprogram_data[ExperimentResults.MEASUREMENTS_PATH]:
+                self.data[(qprogram_name, measurement_name)] = qprogram_data[
+                    f"{ExperimentResults.MEASUREMENTS_PATH}/{measurement_name}/results"
+                ]
+                # Store the dimensions
+                self.dimensions[(qprogram_name, measurement_name)] = qprogram_data[
+                    f"{ExperimentResults.MEASUREMENTS_PATH}/{measurement_name}/results"
+                ].dims
+
+        return self
+
+    def __exit__(self, *args):
+        """Exit the context manager and close the HDF5 file."""
+        if self._file is not None:
+            self._file.close()
+
+    def get(self, qprogram: int | str = 0, measurement: int | str = 0):
```

Comment:
missing return in function definition

Reply:
Nice catch. Writing the return definition, I saw that the dictionary was too complicated, so I created a dataclass `DimensionInfo` that stores the information more organized.